packages/react-reconciler/src/ReactFiberHooks.js

```diff
@@ -259,6 +281,14 @@ function cloneHook(hook: Hook): Hook {
 
     next: null,
   };
+
+  if (__DEV__) {
+    if (!currentHook || currentHook._debugType !== hook._debugType) {
+      warning(false, 'Bad hook order!');
```

Comment:
Perhaps we can improve the error message here. Add a shortened link to [Rules of Hooks ](https://reactjs.org/docs/hooks-rules.html)?